theano/tensor/opt.py

```diff
@@ -5841,3 +5841,26 @@ def local_add_mul_fusion(node):
 def local_grad_clip(node):
     if isinstance(node.op, theano.gradient.GradClip):
         return node.inputs
+
+@register_canonicalize
+@register_stabilize
+@register_specialize
+@gof.local_optimizer([T.Alloc])
+def local_reduce_double_alloc(node):
+    # Alloc(Alloc(m, x, 1), x, y) -> Alloc(m, x, y)
+    if not isinstance(node.op, T.Alloc):
+        return False
+    if not node.inputs[0].owner or not isinstance(node.inputs[0].owner.op, T.Alloc):
+        return False
+    inputs_outer = node.inputs
+    inputs_inner = node.inputs[0].owner.inputs
+    if len(inputs_outer) != len(inputs_inner):
+        return False
```

Comment:
This is too restrictive restriction. inputs_outer must be longer then inputs_inner. Add a test for that case and update the code bellow to handle it.  Mostly, test this case:

 T.alloc(T.alloc(m, y, 1, 1), x, y, z, w)